File: supporter-product-data/src/main/scala/com/gu/lambdas/ProcessSupporterRatePlanItemLambda.scala
package com.gu.lambdas

import com.amazonaws.services.lambda.runtime.Context
import com.gu.conf.ZuoraQuerierConfig
import com.gu.lambdas.ProcessSupporterRatePlanItemLambda.{contributionIds, discountIds}
import com.gu.model.StageConstructors
import com.gu.model.dynamo.SupporterRatePlanItemCodecs._
import com.gu.model.sqs.SqsEvent
import com.gu.model.zuora.response.MinimalZuoraSubscription
import com.gu.monitoring.SafeLogging
import com.gu.okhttp.RequestRunners.configurableFutureRunner
import com.gu.services.{AlarmService, ConfigService, DiscountService, ZuoraSubscriptionService}
import com.gu.supporterdata.model.Stage.{CODE, PROD}
import com.gu.supporterdata.model.{ContributionAmount, Stage, SupporterRatePlanItem}
import com.gu.supporterdata.services.SupporterDataDynamoService
import io.circe.parser._
import io.circe.syntax.EncoderOps

import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.Future
import scala.concurrent.duration.DurationInt
import scala.util.{Failure, Success}

class ProcessSupporterRatePlanItemLambda extends Handler[SqsEvent, Unit] {

  override protected def handlerFuture(input: SqsEvent, context: Context) = {
    logger.info(s"Received ${input.Records.length} records from the queue")
    Future
      .sequence(input.Records.map { record =>
        val subscription = decode[SupporterRatePlanItem](record.body)
        subscription match {
          case Right(item) => ProcessSupporterRatePlanItemLambda.processItem(item)
          case _ =>
            Future.failed(new Exception(s"Couldn't decode a SupporterRatePlanItem with body: ${record.body}"))
        }
      })
      .map(_ => ())
  }
}

object ProcessSupporterRatePlanItemLambda extends SafeLogging {
  val stage = StageConstructors.fromEnvironment
  val config = ConfigService(stage).load
  val dynamoService = SupporterDataDynamoService(stage)
  val contributionIds = ContributionIds.forStage(stage)
  val discountIds = DiscountService(stage).getDiscountProductRatePlanIds.get
  lazy val contributionAmountFetcher = new ContributionAmountFetcher(config)
  lazy val alarmService = AlarmService(stage)

  private def isRecurringContribution(supporterRatePlanItem: SupporterRatePlanItem) =
    contributionIds.contains(supporterRatePlanItem.productRatePlanId)

  private def addAmountIfContribution(supporterRatePlanItem: SupporterRatePlanItem) =
    if (isRecurringContribution(supporterRatePlanItem)) {
      logger.info(
        s"Subscription ${supporterRatePlanItem.subscriptionName} of supporter ${supporterRatePlanItem.identityId} is a recurring contribution",
      )
      contributionAmountFetcher.fetchContributionAmountFromZuora(supporterRatePlanItem)
    } else
      Future.successful(supporterRatePlanItem)

  private def itemIsDiscount(supporterRatePlanItem: SupporterRatePlanItem) =
    discountIds.contains(supporterRatePlanItem.productRatePlanId)

  def processItem(supporterRatePlanItem: SupporterRatePlanItem) = {
    if (itemIsDiscount(supporterRatePlanItem)) {
      logger.info(s"Supporter rate plan item ${supporterRatePlanItem.asJson.spaces2} is a discount")
      Future.successful(())
    } else
      addAmountIfContribution(supporterRatePlanItem).flatMap(writeItemToDynamo)
  }

  private def writeItemToDynamo(supporterRatePlanItem: SupporterRatePlanItem) = {
    logger.info(s"Attempting to write item to Dynamo - ${supporterRatePlanItem.asJson.noSpaces}")

    dynamoService
      .writeItem(supporterRatePlanItem)
      .map { _ =>
        logger.info(s"Successfully wrote item to Dynamo - ${supporterRatePlanItem.asJson.noSpaces}")
        ()
      }
      .recover { case t: Throwable =>
        logger.error(scrub"Error writing item to Dynamo - ${supporterRatePlanItem.asJson.noSpaces}", t)
        alarmService.triggerDynamoWriteAlarm match {
          case Success(_) => Future.successful(())
          case Failure(exception) =>
            logger.error(scrub"Failed to trigger dynamo alarm", exception)
            Future.successful(())
        }
      }
  }
}

class ContributionAmountFetcher(config: ZuoraQuerierConfig) extends SafeLogging {
  lazy val zuoraService = new ZuoraSubscriptionService(config, configurableFutureRunner(60.seconds))

  private def contributionAmountFromZuoraSubscription(subscription: MinimalZuoraSubscription) =
    for {
      contributionRatePlan <- subscription.ratePlans.find(ratePlan => contributionIds.contains(ratePlan.id))
      charges <- contributionRatePlan.ratePlanCharges.headOption
      amount <- charges.price
    } yield ContributionAmount(amount, charges.currency)

  def fetchContributionAmountFromZuora(supporterRatePlanItem: SupporterRatePlanItem): Future[SupporterRatePlanItem] =
    zuoraService
      .getSubscription(supporterRatePlanItem.subscriptionName)
      .map { sub =>
        val contributionAmount = contributionAmountFromZuoraSubscription(sub)
        logger.info(
          s"Contribution amount for supporter ${supporterRatePlanItem.identityId} is ${contributionAmount}",
        )
        supporterRatePlanItem.copy(contributionAmount = contributionAmount)
      }

}

object ContributionIds {
  def forStage(stage: Stage) = stage match {
    case PROD => List("2c92a0fc5aacfadd015ad24db4ff5e97", "2c92a0fc5e1dc084015e37f58c200eea")
    case CODE => List("2c92c0f85a6b134e015a7fcd9f0c7855", "2c92c0f85e2d19af015e3896e824092c")
  }
}

File: supporter-product-data-ts/src/typescript/lambdas/processSupporterRatePlanItemLambda.ts
import type { Handler, SQSEvent } from "aws-lambda";
import { type Stage, stageFromEnvironment } from "../model/stage";
import type {
  ContributionAmount,
  SupporterRatePlanItem,
} from "../model/supporterRatePlanItem";
import { AlarmService } from "../services/alarmService";
import { ConfigService } from "../services/configService";
import { DynamoService } from "../services/dynamoService";
import {
  type MinimalZuoraSubscription,
  ZuoraSubscriptionService,
} from "../services/zuoraSubscriptionService";

const contributionIdsForStage = (stage: Stage): string[] =>
  stage === "PROD"
    ? ["2c92a0fc5aacfadd015ad24db4ff5e97", "2c92a0fc5e1dc084015e37f58c200eea"]
    : ["2c92c0f85a6b134e015a7fcd9f0c7855", "2c92c0f85e2d19af015e3896e824092c"];

const contributionAmountFromZuoraSubscription = (
  subscription: MinimalZuoraSubscription,
  contributionIds: string[]
): ContributionAmount | undefined => {
  const contributionRatePlan = subscription.ratePlans.find((ratePlan) =>
    contributionIds.includes(ratePlan.id)
  );
  const firstCharge = contributionRatePlan?.ratePlanCharges[0];
  if (firstCharge?.price === undefined) {
    return undefined;
  }

  return {
    amount: firstCharge.price,
    currency: firstCharge.currency,
  };
};

interface ProcessItemDeps {
  discountIds: string[];
  contributionIds: string[];
  getSubscription: (
    subscriptionName: string
  ) => Promise<MinimalZuoraSubscription>;
  writeItem: (item: SupporterRatePlanItem) => Promise<void>;
  triggerDynamoWriteAlarm: () => Promise<void>;
}

const buildDeps = async (): Promise<ProcessItemDeps> => {
  const stage = stageFromEnvironment();
  const configService = new ConfigService(stage);
  const config = await configService.loadZuoraConfig();

  const subscriptionService = new ZuoraSubscriptionService(config);
  const dynamoService = new DynamoService(stage);
  const alarmService = new AlarmService(stage);

  return {
    discountIds: config.discountProductRatePlanIds,
    contributionIds: contributionIdsForStage(stage),
    getSubscription: (subscriptionName) =>
      subscriptionService.getSubscription(subscriptionName),
    writeItem: (item) => dynamoService.writeItem(item),
    triggerDynamoWriteAlarm: () => alarmService.triggerDynamoWriteAlarm(),
  };
};

const addContributionAmountIfNeeded = async (
  item: SupporterRatePlanItem,
  deps: Pick<ProcessItemDeps, "contributionIds" | "getSubscription">
): Promise<SupporterRatePlanItem> => {
  if (!deps.contributionIds.includes(item.productRatePlanId)) {
    return item;
  }

  const subscription = await deps.getSubscription(item.subscriptionName);
  const contributionAmount = contributionAmountFromZuoraSubscription(
    subscription,
    deps.contributionIds
  );

  return {
    ...item,
    contributionAmount,
  };
};

export const processItem = async (
  item: SupporterRatePlanItem,
  deps: ProcessItemDeps
): Promise<void> => {
  console.info("Processing supporter rate plan item", {
    subscriptionName: item.subscriptionName,
    identityId: item.identityId,
    productRatePlanId: item.productRatePlanId,
    productRatePlanName: item.productRatePlanName,
    termEndDate: item.termEndDate,
  });

  if (deps.discountIds.includes(item.productRatePlanId)) {
    console.info("Supporter rate plan item is a discount and will be skipped", {
      subscriptionName: item.subscriptionName,
      productRatePlanId: item.productRatePlanId,
    });
    return;
  }

  try {
    const itemWithContribution = await addContributionAmountIfNeeded(
      item,
      deps
    );

    if (itemWithContribution.contributionAmount !== undefined) {
      console.info("Resolved contribution amount", {
        subscriptionName: item.subscriptionName,
        amount: itemWithContribution.contributionAmount.amount,
        currency: itemWithContribution.contributionAmount.currency,
      });
    }

    console.info("Writing item to DynamoDB", {
      subscriptionName: item.subscriptionName,
      identityId: item.identityId,
    });

    await deps.writeItem(itemWithContribution);

    console.info("Successfully wrote item to DynamoDB", {
      subscriptionName: item.subscriptionName,
      identityId: item.identityId,
    });
  } catch (error) {
    console.error("Error writing item to Dynamo", {
      subscriptionName: item.subscriptionName,
      identityId: item.identityId,
      error,
    });
    await deps.triggerDynamoWriteAlarm();
  }
};

const parseItem = (body: string): SupporterRatePlanItem => {
  try {
    return JSON.parse(body) as SupporterRatePlanItem;
  } catch {
    throw new Error(
      `Couldn't decode a SupporterRatePlanItem with body: ${body}`
    );
  }
};

export const processEvent = async (
  event: SQSEvent,
  deps: ProcessItemDeps
): Promise<void> => {
  console.info("Processing SQS event", { recordCount: event.Records.length });

  await Promise.all(
    event.Records.map(async (record) => {
      const item = parseItem(record.body);
      await processItem(item, deps);
    })
  );

  console.info("Finished processing SQS event", {
    recordCount: event.Records.length,
  });
};

export const handler: Handler<SQSEvent, void> = (event) =>
  buildDeps().then((deps) => processEvent(event, deps));
